### DISCUSSION TITLE
Cannot access constant on class-string<T>

### DISCUSSION THREAD

**Discussion by eithed** [spaze/phpstan-disallowed-calls#185] [Q&A] [answered]
Hello!

I've run into the issue:

```
Cannot access constant FOO on class-string<Bar>
```

referenced in https://github.com/spaze/phpstan-disallowed-calls/issues/44

My code is following (well, the gist of it is):

```
/**
 * @return class-string<Bar>
 */
private function getBarClass(string $bar): string
{
    [do things to $bar]

    if (! class_exists($bar) || ! is_a($bar, Bar::class, true)) {
        throw new \Exception("Class does not exist")
    }

    return $bar;
}
```

While I cannot speak of strings and (just) class-strings, if `class-string<T>` is used, then within ClassConstantUsages one could replace

```
$usedOnType = $this->typeResolver->getType($node->class, $scope)
```
with
```
$usedOnType = $this->typeResolver->getType($node->class, $scope)->getObjectTypeOrClassStringObjectType();
```
and then report error on `no` or `maybe`, ie:
```
if ($usedOnType->hasConstant($constant)->yes()) {
   $classNames = [$usedOnType->getConstant($constant)->getDeclaringClass()->getDisplayName()];
} else {
return [
   RuleErrorBuilder::message(sprintf(
      'Cannot access constant %s on %s',
      $constant,
      $usedOnType->describe(VerbosityLevel::getRecommendedLevelByType($usedOnType))
   ))->build(),
];
}
```
as `usedOnType` will be resolved to (correct) Bar, rather than to `{}`. If class-string without class specific is used, current behaviour of alerting about accessing constant on unknown class will be maintained.

I'm happy to create a PR, if you think this is a good idea :)

**Comment by spaze** [comment]
Hey, please create a PR and let's see :-) Thanks!

**Reply by eithed** [reply, marked as answer]
Done! https://github.com/spaze/phpstan-disallowed-calls/pull/186

**Comment by spaze** [comment]
The PR released  in [2.14.1](https://github.com/spaze/phpstan-disallowed-calls/releases/tag/v2.14.1) (replaced 2.14.0 with a quick fix #189)
